src/vsr/journal.zig

```diff
@@ -1356,93 +1355,154 @@ pub fn JournalType(comptime Replica: type, comptime Storage: type) type {
+        ///   other than "torn prepares", which manifest as:
         ///   - the redundant header is valid,
-        ///   - the redundant header is reserved, and/or the op is at least a log cycle behind,
-        ///   - the prepare is corrupt, and
-        /// * there are no faults except for those between `op_checkpoint` and `op_max + 1`,
-        ///   so that we can be sure that the maximum valid op is in fact the maximum.
-        fn recover_torn_prepare(journal: *const Journal, cases: []const *const Case) ?Slot {
+        ///   - the redundant header's op is at least a log cycle behind,
+        ///   - the prepare is corrupt
+        /// * faults may exist outside of (op_max, op_prepare_max]. They have no bearing on the
+        ///   certainty of op_max as they lie between (op_checkpoint, op_max].
+        fn torn_prepares(
+            journal: *const Journal,
+            cases: []const *const Case,
+        ) stdx.BoundedArray(Slot, constants.journal_iops_write_max) {
             const replica: *const Replica = @alignCast(@fieldParentPtr("journal", journal));
 
             assert(journal.status == .recovering);
             assert(journal.dirty.count == slot_count);
             assert(journal.faulty.count == slot_count);
 
-            const op_max = op_maximum_headers_untrusted(replica.cluster, journal.headers_redundant);
-            if (op_max != op_maximum_headers_untrusted(replica.cluster, journal.headers))
-                return null;
-            if (op_max < replica.op_checkpoint()) return null;
-            // We can't assume that the header at `op_max` is a prepare — an empty journal with a
-            // corrupt root prepare (op_max=0) will be repaired later.
-
-            const torn_op = op_max + 1;
-            const torn_slot = journal.slot_for_op(torn_op);
+            const op_max = @max(
+                op_maximum_headers_untrusted(replica.cluster, journal.headers_redundant),
+                op_maximum_headers_untrusted(replica.cluster, journal.headers),
+            );
 
-            const torn_prepare_untrusted = &journal.headers[torn_slot.index];
-            if (torn_prepare_untrusted.valid_checksum()) return null;
-            // The prepare is at least corrupt, possibly torn, but not valid and simply misdirected.
+            // Nothing to truncate - head op is not certain as it must be >= op_checkpoint.
+            if (op_max < replica.op_checkpoint()) return .{};
 
-            const header_untrusted = &journal.headers_redundant[torn_slot.index];
-            const header = header_ok(replica.cluster, torn_slot, header_untrusted) orelse
-                return null;
-            // The redundant header is valid, also for the correct cluster and not misdirected.
+            // Nothing to truncate - replica could not have prepared anything beyond prepare_max.
+            if (op_max == replica.op_prepare_max()) return .{};
```

Comment:
My reasoning here was: our objective here is to truncate torn prepares that were being appended to the _end_ of the WAL, if we are able to discern a certain _end/head_. In this case, since we've found `op_prepare_max` in our journal, it is guaranteed to be certain because this replica couldn't have prepared any op greater than that. Also, since the replica couldn't have been appending any op > op_prepare_max to the end of the WAL, there is nothing to truncate. 

As for the ops _before_ `op_prepare_max`, if they were being prepared before we crashed, they would be categorized as `@B/@C` and fixed using our regular repair.

Thoughts?

Reply:
Mhmmmm, but I get what you're thinking now: we _could_ be repairing old prepares in slots past `op_prepare_max` and could crash while writing those. If we don't truncate those here, that would needlessly land us in `recovering_head`... This means torn prepares don't necessarily have to be "potentially greater" prepares, they could be _any_ prepares after our head. 🤔

Reply:
> we could be repairing old prepares in slots past op_prepare_max and could crash while writing those.

Actually that's not quite what I was thinking.

Suppose two prepares are being prepared concurrently:
- op=X, which is 1MiB, and
- op=X+1, which is 1KiB.

Op=X+1 finishes successfully, but before op=X is done writing we crash, tearing the prepare write.

Reply:
Mhmmmm, but in this case couldn't we just let op=X be fixed using our regular repair protocol?

Reply:
Yes, let me update the comment! 

Also, wanted to get your input about https://github.com/tigerbeetle/tigerbeetle/pull/2345#discussion_r1777409243 as well, which _seems_ like a valid concern to me. If we are repairing some old ops to the right of op_prepare_max and they are torn, we it seems like that could currently land us in .recovering_head since we bail out...

I would have to think about this in terms of a concrete example though, with op_repair_min and op_repair_max, to understand this scenario more clearly. 😄

Reply:
> we could be repairing old prepares in slots past op_prepare_max and could crash while writing those. If we don't truncate those here, that would needlessly land us in recovering_head... This means torn prepares don't necessarily have to be "potentially greater" prepares, they could be any prepares after our head.

Checking my understanding: Would this only impact slots that are also to the left of `op_checkpoint`?

Reply:
Isn't it the other way around? If `op_head == op_prepare_max`, then we _definitely_ know the head op, since we couldn't possibly have prepared anything beyond `op_prepare_max`. (Since `op_prepare_max` is derived from our superblock checkpoint).

Reply:
But could we still be repairing old ops between [op_checkpoint, op_prepare_max]? 🤔

Reply:
Right, but how would that lead to `recovering_head`?

Reply:
Mhmmmm, I was thinking that if the prepares being repaired are torn, they'd be marked as faulty and our `op_head_certain` logic would cause us to go into recovering_head... https://github.com/tigerbeetle/tigerbeetle/blob/f95789c065fd37897b65472b9cf57836771f710f/src/vsr/replica.zig#L5893-L5906

Reply:
Oh wait, `slot_known_range` actually turns out to be an empty range in that case, would `slot_known_range.contains` actually panic here if we have faulty slots or am I missing something obvious? 😮

Reply:
Oh no, scratch that we have the following logic that would prevent that.
https://github.com/tigerbeetle/tigerbeetle/blob/f95789c065fd37897b65472b9cf57836771f710f/src/vsr/replica.zig#L5859-L5869

Reply:
But, this logic _would_ lead us into recovering_head in case there are any faults, right? 🤔

Reply:
Maybe `op_head_certain` should take `op_prepare_max` into account? e.g.:
- Assert that `op_head <= op_prepare_max`, and
- `slot_known_range = slot_op_checkpoint .. slot_op_prepare_max`

Reply:
Mhmmmm you're right, `op_head_certain` could be a good place to handle this! If our head op is op_prepare_max, then we could bail out early since it's guaranteed to be certain (faults in the journal can't affect the certainty in this case)...

Reply:
Took care of this in https://github.com/tigerbeetle/tigerbeetle/pull/2345/commits/95cc35ab166ff11f8a3061a1b7585ffe7522c483. It has been passing continuously running VOPR runs that I've started locally, but I definitely want to let this soak on CFO for a while...

Reply:
Okay, happy about this commit now, ~2.1k passed seeds on my local VOPR run and doing well on the CFO as well!